### ISSUE TITLE
Create TMS processing pipeline

### ISSUE THREAD

**Issue by lossyrob** [hotosm/oam-server#3]
The TMS creation process is kicked off by the TMS Generation web service, and takes as input a list of URIs to the individual images that will be processed.

The imagery will then be processed as follows:
For each image:
- Reproject to EPSG:3857
- Create a VRT from the imagery
- Convert the RGB GeoTiff to PNG

At this stage, we could follow one of two paths, which I will call "static" and "dynamic":

**Static Path:**
A process is kicked off which reads each TMS tile at each zoom level. These images are saved off into an S3 bucket at the appropriate {z}/{x}/{y}.png subkey so that the S3 bucket operates as a TMS service addressing scheme.

**Dynamic Path:**
All upfront processing is completed; the TMS service points to a server address which has an endpoint that takes the TMS request, translates it into a windowed tile request out of the VRT, and returns the PNG.

There is also the potential of a **Hybrid Path**, which perhaps stores the static files from some zoom levels and dynamically renders the tiles for zoom levels which are not created upfront.

This pipeline could be a generalization of this code: https://github.com/stamen/swfr/blob/master/cgiar-csi-srtm.js

**Comment by lossyrob** [comment]
My recommendation after talking with Seth is that we code the _Dynamic Path_, test for performance. If it turns out that is too slow, we still have coded the necessary components for the static path, its just a matter of moving when that code will run.

**Comment by cgiovando** [comment]
:+1: on **Hybrid** solution, but we need to test how seamless would be going from pre-seeded tiles to dynamically generated if all the requests don't pass first through a Tilestache-like endpoint.

@mojodna  mentioned some S3 property that lets trigger processes if a file is not found. That would be nice and allow for having tile services primarily hosted on S3, but will also make OAM AWS-dependent.

Unless we decide to always entirely pre-seed all tile services for every service hosted on OAM, then the **Dynamic** option is probably what we need to go with.

**Comment by cgiovando** [comment]
Since _"TMS"_ usually refers to the OSGeo tile service specification, I would recommend using something generic like _"tile service"_ or _"tile map service"_ spelling to avoid confusion until we close #5

**Comment by sethfitz** [comment]
> Convert the RGB GeoTiff to PNG

I don't think this is necessary.

**Comment by sethfitz** [comment]
> @mojodna mentioned some S3 property that lets trigger processes if a file is not found. That would be nice and allow for having tile services primarily hosted on S3, but will also make OAM AWS-dependent.

Not necessarily.  If we require that clients follow redirects, this approach will work with S3 and non-AWS instances.  Non-AWS instances then have the choice of whether to redirect 404s to a secondary server or whether to pass them through to a render instance internally.

Here's how to do it on S3:

https://docs.aws.amazon.com/AmazonS3/latest/dev/HowDoIWebsiteConfiguration.html

E.g.:

``` xml
<RoutingRules>
    <RoutingRule>
        <Condition>
            <!-- 403 not 404, as default bucket configurations return 403 Forbidden on non-existent (as well as actually forbidden) objects -->
            <HttpErrorCodeReturnedEquals>403</HttpErrorCodeReturnedEquals>
        </Condition>
        <Redirect>
            <HostName>www.example.com</HostName>
            <ReplaceKeyPrefixWith>PutSomePrefixHere</ReplaceKeyPrefixWith>
            <HttpRedirectCode>303</HttpRedirectCode>
        </Redirect>
    </RoutingRule>
</RoutingRules>
```

**Comment by sethfitz** [comment]
Another way to look at the _hybrid_ path is as the _dynamic_ path, but with external caching that continues to work even if the dynamic rendering instances are down.  Bits of the approach are laid out in https://speakerdeck.com/mojodna/running-your-own-rendering-infrastructure-foss4g-2014-edition

**Comment by lossyrob** [comment]
> I don't think this is necessary.

We'd need to convert it from tiff to display on web maps, correct?

**Comment by cgiovando** [comment]
With regards to  the stack for a more dynamic implementation, wondering how Tilestream + MBTiles compares to mod_onearth + MRF and tilestache/geowebcache + plain PNG caches.

https://www.mapbox.com/blog/building-mapbox-fast-map-hosting-stack/
https://github.com/mapbox/tilestream

https://github.com/nasa-gibs/onearth/tree/master/src/mod_onearth
https://github.com/nasa-gibs/mrf/blob/master/spec/mrf_spec.md

https://github.com/TileStache/TileStache
https://github.com/GeoWebCache/geowebcache

**Comment by sethfitz** [comment]
I'm partial to a CDN (Fastly, in my case), backed by S3 and [tessera](https://github.com/mojodna/tessera) w/ tilelive-mapnik (or [tilelive-raster](https://github.com/mojodna/tilelive-raster)), in part because the Node pieces are the same that Mapbox has been putting lots of effort into.

**Comment by cgiovando** [comment]
How does tessera handle requests for tiles already existing in cache and tile expiration?  
Can it pre-seed entire services?  
Does it need to go through GDAL to proxy cache a WMS?  
Can it write and read tile caches directly to S3?

**Comment by sethfitz** [comment]
> Does it need to go through GDAL to proxy cache a WMS?

At the moment, yes, though there are a few Node.js libraries that would enable us to remove the GDAL dependency (and thus the inherent thread and concurrency limits).  E.g. https://github.com/JasonSanford/tileify-wms

> How does tessera handle requests for tiles already existing in cache and tile expiration?
> Can it write and read tile caches directly to S3?

It doesn't.  It was a very intentional choice to omit any notion of caching (beyond a small in-memory cache) and punt that to a different architectural layer (based on my experience with TileStache and maps.stamen.com).

tessera itself is as minimal as possible, leaning on tilelive modules (or composed sets of tilelive modules) to achieve anything.

tessera assumes that you have other methods for caching at your disposal, either a CDN or a local proxy server.  We use Fastly, which has the ability to distinguish between client caching behavior (`Cache-Control`) and server caching behavior (`Surrogate-Control`).

Since CDNs operate multiple instances that can't share the same cache, we use a reverse proxy ([tp](https://github.com/mojodna/tp) on a single Heroku dyno) to ensure that newly rendered tiles find their way into S3.  I'm not 100% content with the way this works (especially since it represents a single point of failure), but it's been serving tile.stamen.com for about a year (and Map Stack for longer) without any hiccups.

tessera also includes the ability to customize the headers returned for individual tiles, which allows for [Surrogate Keys](https://www.fastly.com/blog/surrogate-keys-part-1/) to be defined so that entire blocks of tiles can be invalidated in one go.

> Can it pre-seed entire services?

Not by itself.  [tl](https://github.com/mojodna/tl) (which also relies on the tilelive ecosystem) can be used in this mode, either by rendering locally and writing to some other source or by requesting swaths of tiles from a fully configured (i.e. caching) stack and writing them into a null sink.  We've also used [kevlar](https://github.com/stamen/kevlar) for this purpose, though that was using existing traffic to generate load against a new set of tile servers (we did this for a couple days to pre-seed our Toner updates with more highly requested content).

Fundamentally, tessera is modular where tilestache is monolithic. The advantage of modular is the existence of various tilelive modules, some of which are used in Mapbox's client tools and rendering infrastructure.
